File: 12_integer_to_roman/integer_to_roman.java
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

class Solution {
    public String intToRoman(int num) {
        Map<Integer, String> table = new LinkedHashMap<Integer, String>();
        table.put(1000, "M");
        table.put(900, "CM");
        table.put(500, "D");
        table.put(400, "CD");
        table.put(100, "C");
        table.put(90, "XC");
        table.put(50, "L");
        table.put(40, "XL");
        table.put(10, "X");
        table.put(9, "IX");
        table.put(5, "V");
        table.put(4, "IV");
        table.put(1, "I");
        String result = "";
        while (num > 0) {
            for (Map.Entry<Integer, String> entry: table.entrySet()) {
                Integer key =entry.getKey();
                if (num >= key) {
                    num -= key;
                    result += entry.getValue();
                    break;
                }
            }
        }
        return result;
    }
}

File: 12_integer_to_roman/integer_to_roman.c
char * intToRoman(int num){
    // 使用两个数组模拟哈希表
    int keys[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
    char values[13][3] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
    char *p = (char *)calloc(16, sizeof(char));  // 由于num < 3999 因此16个字符长度已经足够
    for (int i = 0; i < 13; i++) {
        while (num >= keys[i]) {
            strcat(p, values[i]);
            num -= keys[i];
        }
    }
    return p;
}